lib/bandit/websocket/handler.ex

```diff
@@ -1,31 +1,32 @@
 defmodule Bandit.WebSocket.Handler do
   @moduledoc false
-  # A WebSocket handler conforming to RFC6455. As a Handler, this works with raw
-  # ThousandIsland Sockets and supports HTTP/1.1-sourced WebSocket connections
+  # A WebSocket handler conforming to RFC6455, structured as a ThousandIsland.Handler
 
   use ThousandIsland.Handler
 
   alias Bandit.WebSocket.{Connection, Frame}
 
   @impl ThousandIsland.Handler
   def handle_connection(socket, state) do
-    connection = Connection.init(state.sock)
+    {sock, sock_opts, connection_opts} = state.upgrade_opts
+
+    state =
+      state
+      |> Map.take([:handler_module])
+      |> Map.put(:buffer, <<>>)
+
+    case Connection.init(sock, sock_opts, socket) do
+      {:continue, connection} ->
+        case Keyword.get(connection_opts, :timeout) do
+          nil -> {:continue, Map.put(state, :connection, connection)}
+          timeout -> {:continue, Map.put(state, :connection, connection), {:persistent, timeout}}
+        end
 
-    Connection.handle_connection(state.conn, socket, connection)
-    |> case do
-      {:continue, connection} -> {:continue, update_state(state, connection)}
-      {:continue, connection, timeout} -> {:continue, update_state(state, connection), timeout}
-      {:close, _} -> {:close, state}
+      {:error, reason, connection} ->
+        {:error, reason, state |> Map.put(:connection, connection)}
```

Comment:
```suggestion
        {:error, reason, Map.put(state, :connection, connection)}
```
to be consistent with the previous cases